### ISSUE TITLE
Block not defined in blockToXML

### ISSUE THREAD

**Issue by paulkaplan** [LLK/scratch-vm#589]
During our internal playtest, many projects eventually stopped being able to switch sprites, the error in the console was 

```
lib.min.js:6 Uncaught TypeError: Cannot read property 'shadow' of undefined
    at Blocks.blockToXML (https://llk.github.io/scratch-gui/gui.js:8668:32)
    at Blocks.blockToXML (https://llk.github.io/scratch-gui/gui.js:8686:43)
    at Blocks.blockToXML (https://llk.github.io/scratch-gui/gui.js:8703:46)
    at Blocks.blockToXML (https://llk.github.io/scratch-gui/gui.js:8703:46)
    at Blocks.blockToXML (https://llk.github.io/scratch-gui/gui.js:8682:43)
    at https://llk.github.io/scratch-gui/gui.js:8652:31
    at Array.map (native)
    at Blocks.toXML (https://llk.github.io/scratch-gui/gui.js:8651:34)
    at VirtualMachine.emitWorkspaceUpdate (https://llk.github.io/scratch-gui/gui.js:22612:128)
    at VirtualMachine.setEditingTarget (https://llk.github.io/scratch-gui/gui.js:22568:22)
```

I'm not sure why it is happening, but when `blockToXML` is called, some block is not defined. 

Need to investigate and create a repro test case, as I don't have a way to repro beyond just using the app for a while.

**Comment by jwzimmer-zz** [comment]
Further details from the playtest:

### Expected Behavior

Clicking on a sprite should bring up the blocks associated with that sprite.

### Actual Behavior

Reported by Shruti and others during the Catch Game internal playtest (6/2/17):
> I added the script for the falling object (hat) and then the catcher (tera), and then when I clicked on Hat sprite, the script is no more there, and the the script in tera is showing inthe hat sprite. It happened twice.

Several people ran into this, including @chrisgarrity. It looked like the scripts from one sprite were overwriting the scripts for another, since when you clicked between or changed the script for sprite A, it would show up for sprite B as well. Apparently, the scripts for sprite B were likely still there, just not being returned / visible in the scripting area.

### Operating System and Browser

_Mac OS, Chrome_

**Comment by paulkaplan** [comment]
So far this is the simplest repro I can figure out. Still kind of baffled by this though.
![block-to-xml-error](https://user-images.githubusercontent.com/654102/26847266-6059624a-4aca-11e7-9515-7d7dc9e7bc2d.gif)

Steps:
1. Add a block with a shadow field (e.g. "move __ steps")
2. Add a reporter into that shadow field (e.g. "x position")
3. Duplicate the block by right clicking
4. Connect the two blocks
5. Delete the top block
6. Switch sprites and switch back

**Comment by paulkaplan** [comment]
@rschamp I think I've got a handle on this now.

It seems that scratch-blocks allows duplicate IDs for obscured shadow blocks ([confirmed here](https://github.com/LLK/scratch-blocks/issues/926)). When a block with a duplicated obscured shadow is deleted, the VM deletes the shadow block from its block cache. 

It would seem from the [code comment](https://github.com/LLK/scratch-vm/blob/develop/src/engine/blocks.js#L232-L233) that this behavior is not intended, and obscured shadow block shouldn't be deleted. However it [happens anyway here](https://github.com/LLK/scratch-vm/blob/develop/src/engine/blocks.js#L401-L405). Removing those lines solves the bug.

TL;DR it is not safe to delete obscured shadow blocks because they may live in multiple places. 

I feel that this should be changed in scratch-blocks, because the root cause is a block ID collision. Thoughts /cc @thisandagain ?

**Comment by thisandagain** [comment]
Thanks for tracking this down @paulkaplan ! I agree that we should resolve the root cause in `scratch-blocks` after consulting with the Google team.

/cc @picklesrus @rachel-fenichel

**Comment by paulkaplan** [comment]
fixed by https://github.com/LLK/scratch-blocks/pull/936
